assets/src/edit-story/elements/text/display.js

```diff
@@ -143,7 +131,7 @@ function TextDisplay({
             <BackgroundSpan
               {...props}
               dangerouslySetInnerHTML={{
-                __html: draftMarkupToContent(content, bold),
+                __html: content,
```

Comment:
There's a value, I believe, in having `<strong>` in the markup. It's a pretty reasonable flow (you can try it in Google Slides).

1. Select the text as a frame. Toggle "Bold".
2. Go to text edit mode.
3. Select partial text and press <kbd>cmd+b</kbd>.
4. Notice that the selected text is no longer bold.

I think @miina did some initial research here, and I think I agree with her on this: the toggling buttons (B/I/etc) should modify markup. Perhaps the problem is different here and we are not taking that approach far enough? E.g. maybe we should do the operations on the `content` field only for these fields and remove `element.bold`, `element.fontStyle == italic`, etc?

And I think if we went down this path, then the other code changes here would be unnecessary. I think it could be pretty straightforward to update markup via `<template>` tag and do it relatively optimally. E.g. after a cmd+b, a hypothetical `<strong><string>x</strong> y</strong>` markup should be fairly easily optimizable to `<strong>x y</strong>`. Though that might be even unnecessary at the first pass, as long as draft.js can accept such a markup.

Reply:
Related: https://github.com/google/web-stories-wp/issues/413
Note that the toggles should (eventually) behave the same way as shortcuts: when selecting just part of the text and using toggles, it would be applied only on that part of the text, not the full content. Since currently, we don't support marking just part of the text and toggling, it's pretty simple -- just wrapping the content fully into `strong` (etc.), however, the idea was to remove `element.bold` (etc.) eventually indeed.

So yes, agreed here with Dima that perhaps we could get rid of the element attributes right away and operate on the content only, even if still not handling partial markup change via toggles.

Reply:
So just to clarify, are you both saying to change the italic/bold/underline system implemented currently? I mean, currently the italic and underline are applying the respective CSS styles (e.g. `font-style` and `text-decoration`) in the content parent element, no matter which markup is inside. The proposal of this PR is to create a proper fix for the bold button action given the current standard. That being said, are you both inclined to use tags instead of CSS for bold/italic/underline? Should I start this refactoring using this bugfix PR or another one?

Reply:
BTW, I [even asked](https://xwp.slack.com/archives/CCCAUAH6F/p1586205662471800) about what approach to use before creating this draft once I see these mix of patterns being used :) Let me know if you have additional thoughts after read that report.

Reply:
@obetomuniz Sorry for not reacting on that question, please feel free to tag specifically the next time if you see that no one has answered in Slack -- things might get lost/forgotten easily.

IMO, ideally, we'd move more towards what DraftJS is doing and work with the content directly since the behavior of the toggles should match the shortcodes eventually.

Not sure if this specific issue is a blocker for alpha (cc @swissspidy ?) -- if it's not, then it would be great to already fix it moving towards making DraftJS and the toggles work better together.

I know that @dvoytenko was at some point looking into Toggles + partial selection as well so he might have better insights here.

Also, Gutenberg has the partial selection logic implemented as well, might also be good for inspiration (even if we're currently not implementing the partial selection yet via Toggles).

Thoughts?

Reply:
I think it's definitely a goal to ensure that whole-text vs selection Bold/Italic/Underline work correctly. This is not even the future: it's broken right now. E.g.
 1. Take a text element
 2. Click on the underline button in the design panel. Observe it's all italic.
 3. Edit text.
 4. Select part of text and click cmd+I. Observe that italic is not toggled off.

I think this needs to be fixed before other things in this category, but they will drastically change how we look at the resulting markup. And with DraftJS, I believe, the only way here is to toggle these decorations on the markup itself, not container styles.

Reply:
> 2. Click on the underline button in the design panel. Observe it's all italic.

That _does_ sound like a bug...  🤔  😉.

But yes, if we changed the B / I / U buttons to wrap/unwrap the whole element in a `<strong>`, `<em>`, and `<u>` respectively, that would probably work.

Then, when it comes to setting font weight from the dropdown however, we'd need to do some extra magic. Something like:

1. If you select a non-bold font (weight<700), add a CSS rule that the base font weight of the element is X (`.element { font-weight: 300; }`).
1. If you select a bold font (weight>=700), wrap the element in a `<strong>` tag and add a CSS rule that sets the font weight for strong tags to X (`.element strong { font-weight: 900; }`).

But then what happens when you enter edit mode and makes part of the text not bold with respect to the font weight dropdown? Let's say this happens:

1. Create a text element with the text "123" ➡️ text content: `'123'`
2. Select font weight _Heavy_ (900) from the dropdown  ➡️ text content: `'<strong>123</strong>'`.
3. Edit the text field, select the first two letters and click MOD+b to _unbold_ them ➡️ text content: `'12<strong>3</strong>'`.
4. What does the font weight dropdown display now? Still _Heavy_ (900) even though most of the text is _Regular_ (400)?
5. How do you change the weight of the regular weight text? What if you want bold to be 900 and regular weight to be 300?

It's a pretty complex issue - a single font weight drop down does not really help us out here. As soon as you mix bold and regular text, the font weight dropdown kind of just has to say 🤷.

Or we'd need to have 2 dropdowns - one for setting "regular" font weight, another for setting "bold" font weight - and then you can set those to whatever makes sense. But that's a stupidly complex UX all of a sudden.

/cc @pbakaus

Reply:
1. We should definitely not have two dropdowns ;)
2. When there are multiple weights used in a text element and there's no text selection, the font-weight drop down should say "Multiple", and when you click into the dropdown there should be a dash before the used fonts (I think this is analogous to almost all other editors)
3. When there you've selected a portion of the text, the same rule as above applies: If there are multiple weights in the selection do the above, if there's only one weight found in the selection, display that weight.
4. "Unbolding" a text selection with CMD+B always returns to 400 weight, with a single exception: If the rest of the non-bold text in the text element is 300, we drop to 300.
5. "Bolding" Text with CMD+B always does a 700 weight.
6. If you select a portion of text that includes a combination of weights, CMD+B will bolden everything in the selection to either 700 or the biggest weight in the previous selection.
7. If you select a portion of text that includes a single or multiple levels of boldness and no normal/light text (>400), CMD+B unbolds everything to 300/400 (based on rule in point 4)
8. If you have any kind of text selection and change the weight via dropdown, everything in the selection switches to that weight.

I can't think of any other edge cases right now, but feel free to throw them at me if you find more. I think if you implement the above list exactly, you should be in a pretty good state.

(@o-fernandez if you have cycles to spare, this would actually be a really nice detailed problem to document in the spec under the font variant/weight section of the design panel spec :) )

Reply:
Here's what I would like to propose as a solution here: https://github.com/google/web-stories-wp/pull/1151

@pbakaus note that I propose a somewhat different approach to bolds over 700.

Reply:
@dvoytenko in your approach, you're essentially detaching the "b" button from font-variant. It's an interesting proposal, but has pros and cons as well - tbh, mostly cons in my mind:

1. If you have a font with weight 700 for the whole text element, is the "B" active/toggled? If so, if you do CMD+B on a selection to go to 800, isn't that weird because the B was already toggled? As user, I would expect the reverse to happen if the "B" is visually active (unbold).
2. If we completely detach "B" from font variant, it means that you can click "B", and it won't actually change the font-variant, and vice versa. Seems slightly strange as well.

I think it's time for @samitron7 to chime in :)

Reply:
For situations like these, I generally try to adapt a common pattern that other design tools are already using (the longer the history of the tool, the better). We don't want to break or create a new pattern.

**Figma/Sketch:**
1. I have a font with weight 700 (BOLD) on the page
2. I CMD+B, this should toggle the font to REGULAR
3. Then I change the font to [semi-bold, extra-bold etc..]
4. I CMD+B, this will toggle the font to REGULAR
5. I CMD+B again. This will change the font to BOLD (weight 700)

**Keynote:**
1-4 are the same
5. CMD+B again will toggle the font back to the style found in step 3.

I prefer the keynote way because:
1. First time I CMD+B a text element, I always expect it to be weight 700 (BOLD). If I want any other bold styling (ie semi-bold, black etc), I would go to the dropdown
2. If I made the effort to go to the dropdown to change the font to Black and then hit CMD+B twice, I want that Black style to come back.  The assumption here is that if you're doing a CMD+B twice in a row, you'll using it like a toggle state.

Reply:
Ok, @pbakaus @samitron7 what would really help here, is if we can decide which font fields apply to a part of text and which to the whole text element. We have fontFamily, fontSize, fontWeight, lineHeight, letterSpacing, alignment, bold, italic, underline, color, fill.

From reading everything above, I can surmise that `lineHeight`, `alignment`, and `fill` are element-level. I assume that means that `fontSize` is probably too? Can you define it for all the rest of properties and/or correct if my assumptions?

Reply:
To be safe, yes, let's use the list above, and also font-family and height. So for the time being, these are the global (element level) properties:

- line height
- alignment
- fill
- text color
- padding
- font family
- font size

but @o-fernandez is also working it into the spec as I write (thanks!).